### ISSUE TITLE
Zeroconf.ZeroconfHost.AddService (Zeroconf.IService service) Null Key Exception - iOS, Bug

### ISSUE THREAD

**Issue by twofingerrightclick** [novotnyllc/Zeroconf#244]
Using latest stable Zeroconf 3.6.11
I am trying to get an iOS Xamarin project working for iOS 14.5+ (previously worked on iOS versions prior). It only needs to search for a single service and so I have followed the instructions for adding `NSBonjourServices` in the `Info.plist`:
```
<key>NSLocalNetworkUsageDescription</key>
	<string>Looking for printers with mDNS/Bonjour services</string>
<key>NSBonjourServices</key>
	<array>
		<string>_printer._tcp</string>
	</array>
```

When calling  `ZeroconfResolver.ResolveAsync("_printer._tcp.local.")` the following exception bubbles up from Zeroconf.ZeroconfHost.AddService (Zeroconf.IService service): 

> {System.ArgumentNullException: Value cannot be null.Parameter name: key  at System.Collections.Generic.Dictionary`2[TKey,TValue].TryInsert (TKey key, TValue value, System.Collections.Generic.InsertionBehavior behavior) [0x00008] in <35fb05e83cf64d249b5a7be1a6f7838a>:0   at System.Collections.Generic.Dictionary`2[TKey,TValue].set_Item (TKey key, TValue value) [0x00000] in <35fb05e83cf64d249b5a7be1a6f7838a>:0   at Zeroconf.ZeroconfHost.AddService (Zeroconf.IService service) [0x00000] in /_/Zeroconf/ZeroconfRecord.cs:193   at Zeroconf.BonjourBrowser.RefreshZeroconfHostDict () [0x00121] in /_/Zeroconf/BonjourBrowser.cs:436   at Zeroconf.BonjourBrowser.ReturnZeroconfHostResults () [0x00028] in /_/Zeroconf/BonjourBrowser.cs:361   at Zeroconf.ZeroconfNetServiceBrowser.ResolveAsync (Zeroconf.ResolveOptions options, System.Action`1[T] callback, System.Threading.CancellationToken cancellationToken, System.Net.NetworkInformation.NetworkInterface[] netInterfacesToSendRequestOn) [0x0010b] in /_/Zeroconf/ZeroconfNetServiceBrowser.cs:39   at Zeroconf.ZeroconfResolver.ResolveAsync (Zeroconf.ResolveOptions options, System.Action`1[T] callback, System.Threading.CancellationToken cancellationToken, System.Net.NetworkInformation.NetworkInterface[] netInterfacesToSendRequestOn) [0x000e0] in /_/Zeroconf/ZeroconfResolver.Async.cs:105   at Zeroconf.ZeroconfResolver.ResolveAsync (System.Collections.Generic.IEnumerable`1[T] protocols, System.TimeSpan scanTime, System.Int32 retries, System.Int32 retryDelayMilliseconds, System.Action`1[T] callback, System.Threading.CancellationToken cancellationToken...

`ZeroconfResolver.IsiOSWorkaroundEnabled` resolves to true.

I am trying to run the demo iOS project included in the repo to confirm similar behavior, but I currently can get it working... it builds and deploys, but is just a black screen.

**Comment by twofingerrightclick** [comment]
So I cloned the repo and added the ZeroConf project instead of the nuget package... and no exception. So the nuget package  3.6.11 has the bug.

**Comment by twofingerrightclick** [comment]
So I cloned the repo to use the Zeroconf proj instead of the nuget package 3.6.11 and now there is no longer an exception that bubbles up. Instead I can see in the Zeroconf debug output that there is a System.Net.Sockets.SocketException (0x80004005): No route to host error"

> [0:] AsyncLock.LockAsync called by: ResolveInternal in file: D:\repoD\IntelliTect\Itron\HAN-Meter-Reader-Prototype\src\MobileApp\Zeroconf\ZeroconfResolver.cs : 35
> [0:] Looking for _printer._tcp.local. with scantime 00:00:02
> [0:] Scanning on iface en0, idx 3, IP: 10.10.8.138
> [0:] Attempting to bind to 0.0.0.0:5353 on adapter en0
> [0:] Bound to 0.0.0.0:5353
> [0:] Bound to multicast address
> [0:] About to send on iface en0
> [0:] Scanning on iface en2, idx 5, IP: 169.254.4.100
> [0:] Attempting to bind to 0.0.0.0:5353 on adapter en2
> [0:] Bound to 0.0.0.0:5353
> [0:] Execption with network request, IP 10.10.8.138
> : System.Net.Sockets.SocketException (0x80004005): No route to host
>   at System.Net.Sockets.SocketAsyncResult.CheckIfThrowDelayedException () [0x0003b] in /Library/Frameworks/Xamarin.iOS.framework/Versions/Current/src/Xamarin.iOS/mcs/class/System/System.Net.Sockets/SocketAsyncResult.cs:134 
>   at System.Net.Sockets.Socket.EndSendTo (System.IAsyncResult asyncResult) [0x0002c] in /Library/Frameworks/Xamarin.iOS.framework/Versions/Current/src/Xamarin.iOS/mcs/class/System/System.Net.Sockets/Socket.cs:2393 
>   at System.Net.Sockets.UdpClient.EndSend (System.IAsyncResult asyncResult) [0x0002e] in /Library/Frameworks/Xamarin.iOS.framework/Versions/Current/src/Xamarin.iOS/mcs/class/referencesource/System/net/System/Net/Sockets/UDPClient.cs:697 
>   at System.Threading.Tasks.TaskFactory`1[TResult].FromAsyncCoreLogic (System.IAsyncResult iar, System.Func`2[T,TResult] endFunction, System.Action`1[T] endAction, System.Threading.Tasks.Task`1[TResult] promise, System.Boolean requiresSynchronization) [0x0000f] in /Library/Frameworks/Xamarin.iOS.framework/Versions/Current/src/Xamarin.iOS/external/corert/src/System.Private.CoreLib/src/System/Threading/Tasks/FutureFactory.cs:534 
> --- End of stack trace from previous location where exception was thrown ---

**Comment by twofingerrightclick** [comment]
So in the end this looks like https://github.com/novotnyllc/Zeroconf/issues/195 as I am using the latest XCode 13.4.1

**Comment by twofingerrightclick** [comment]
So the exception comes from here: https://github.com/novotnyllc/Zeroconf/blob/238df25a26d2b0f214d7eee20f3d6a1e105ee9b8/Zeroconf/ZeroconfRecord.cs#L193
The ServiceName on the Service object is not being set (is null) as part of the iOS pipeline of the ResolveAsync. I will see if I can figure out where the ServiceName is supposed to be set, but isn't - Nullable enabled would be helpful here.

**Comment by twofingerrightclick** [comment]
I am thinking it should have been set about here in the BonjourBrowser.cs https://github.com/novotnyllc/Zeroconf/blob/238df25a26d2b0f214d7eee20f3d6a1e105ee9b8/Zeroconf/BonjourBrowser.cs#L399-L402

**Comment by twofingerrightclick** [comment]
@clairernovotny Please review https://github.com/novotnyllc/Zeroconf/pull/246

**Comment by clairernovotny** [comment]
Fixed when #246 was merged
